README.md
# sgminer


## Introduction

This is a multi-threaded multi-pool GPU miner with ATI GPU monitoring,
(over)clocking and fanspeed support for scrypt-based cryptocurrency. It is
based on cgminer by Con Kolivas (ckolivas), which is in turn based on
cpuminer by Jeff Garzik (jgarzik).

**releases**: https://github.com/sgminer-dev/sgminer/releases

**git tree**: https://github.com/sgminer-dev/sgminer

**bugtracker**: https://github.com/sgminer-dev/sgminer/issues

**irc**: `#sgminer` and `#sgminer-dev` on freenode

**mailing lists**: https://sourceforge.net/p/sgminer/mailman/

License: GPLv3.  See `COPYING` for details.


## Documentation

Documentation is available in directory `doc`. It is organised by topics:

* `API` for the RPC API specification;
* `configuration.md` for (largely incomplete) detailed information on all
  configuration options;
* `FAQ.md` for frequently asked questions;
* `GPU` for semi-obsolete information on GPU configuration options and mining
  SHA256d-based coins;
* `kernel.md` for OpenCL kernel-related information, including development
  procedure;
* `MINING.md` for how to find the right balance in GPU configuration to mine
  Scrypt-based coins efficiently;
* `windows-build.txt` for information on how to build on Windows.

Note that **most of the documentation is outdated or incomplete**. If
you want to contribute, fork this repository, update as needed, and
submit a pull request.


## Building

### Dependencies

Mandatory:

* [curl dev library](http://curl.haxx.se/libcurl/) - `libcurl4-openssl-dev` on Debian
* [pkg-config](http://www.freedesktop.org/wiki/Software/pkg-config)
* [libtool](http://www.gnu.org/software/libtool/)
* [AMD APP SDK](http://developer.amd.com/tools-and-sdks/heterogeneous-computing/amd-accelerated-parallel-processing-app-sdk/downloads/)	- available under various names as a package on different GNU/Linux distributions

Optional:

* curses dev library - `libncurses5-dev` on Debian or `libpdcurses` on WIN32, for text user interface
* [AMD ADL SDK](http://developer.amd.com/tools-and-sdks/graphics-development/display-library-adl-sdk/) - version 6, required for ATI GPU monitoring & clocking

If building from git:

* autoconf
* automake

sgminer-specific configuration options:

    --disable-adl           Override detection and disable building with adl
	--disable-adl-checks
    --without-curses        Do not compile support for curses TUI

#### Debian Example

    apt-get install libcurl4-openssl-dev pkg-config libtool libncurses5-dev
AMD APP SDK and AMD ADL SDK must be downloaded from the amd websites.

### *nix build instructions

If needed, place include headers (`*.h` files) from `ADL_SDK_*<VERSION>*.zip` in `sgminer/ADL_SDK`.

Then:

    git submodule init
    git submodule update
    autoreconf -i
    CFLAGS="-O2 -Wall -march=native" ./configure <options>
    make

To compile a version that can be used accross machines, remove
`-march=native`.

To compile a debug version, replace `-O2` with `-ggdb`.

Systemwide installation is optional. You may run `sgminer` from the build
directory directly, or `make install` if you wish to install
`sgminer` to a system location or a location you specified with `--prefix`.

### Windows build instructions

See `doc/windows-build.txt` for MinGW compilation and cross-compiation,
`doc/cygwin-build.txt` for building using Cygwin, or use the provided
`winbuild` Microsoft Visual Studio project (tested on MSVS2010), with
instructions in `winbuild/README.txt`.


## Basic Usage

**WARNING**: documentation below this point has not been updated since the
fork.

After saving configuration from the menu, you do not need to give sgminer
any arguments and it will load your configuration.

Any configuration file may also contain a single

    "include" : "filename"

to recursively include another configuration file.

Writing the configuration will save all settings from all files in the
output.

Single pool:

sgminer -o http://pool:port -u username -p password

Multiple pools:

sgminer -o http://pool1:port -u pool1username -p pool1password -o http://pool2:port -u pool2usernmae -p pool2password

Single pool with a standard http proxy, regular desktop:

sgminer -o "http:proxy:port|http://pool:port" -u username -p password

Single pool with a socks5 proxy, regular desktop:

sgminer -o "socks5:proxy:port|http://pool:port" -u username -p password

Single pool with stratum protocol support:

sgminer -o stratum+tcp://pool:port -u username -p password

The list of proxy types are:
 http:    standard http 1.1 proxy
 http0:   http 1.0 proxy
 socks4:  socks4 proxy
 socks5:  socks5 proxy
 socks4a: socks4a proxy
 socks5h: socks5 proxy using a hostname

If you compile sgminer with a version of CURL before 7.19.4 then some of
the above will not be available. All are available since CURL version
7.19.4.

If you specify the --socks-proxy option to sgminer, it will only be
applied to all pools that don't specify their own proxy setting like
above.

For more advanced usage , run `sgminer --help`.

See `doc/GPU` for more information regarding GPU mining and
`doc/SCRYPT` for more information regarding Scrypt mining.


## Runtime usage

The following options are available while running with a single keypress:

[P]ool management [G]PU management [S]ettings [D]isplay options [Q]uit

P gives you:

Current pool management strategy: Failover
[F]ailover only disabled
[A]dd pool [R]emove pool [D]isable pool [E]nable pool
[C]hange management strategy [S]witch pool [I]nformation


S gives you:

[Q]ueue: 1
[S]cantime: 60
[E]xpiry: 120
[W]rite config file
[C]gminer restart


D gives you:

[N]ormal [C]lear [S]ilent mode (disable all output)
[D]ebug:off
[P]er-device:off
[Q]uiet:off
[V]erbose:off
[R]PC debug:off
[W]orkTime details:off
co[M]pact: off
[L]og interval:5


Q quits the application.


G gives you something like:

GPU 0: [124.2 / 191.3 Mh/s] [A:77  R:33  HW:0  U:1.73/m  WU 1.73/m]
Temp: 67.0 C
Fan Speed: 35% (2500 RPM)
Engine Clock: 960 MHz
Memory Clock: 480 Mhz
Vddc: 1.200 V
Activity: 93%
Powertune: 0%
Last initialised: [2011-09-06 12:03:56]
Thread 0: 62.4 Mh/s Enabled ALIVE
Thread 1: 60.2 Mh/s Enabled ALIVE

[E]nable [D]isable [R]estart GPU [C]hange settings
Or press any other key to continue


The running log shows output like this:

 [2012-10-12 18:02:20] Accepted f0c05469 Diff 1/1 GPU 0 pool 1
 [2012-10-12 18:02:22] Accepted 218ac982 Diff 7/1 GPU 1 pool 1
 [2012-10-12 18:02:23] Accepted d8300795 Diff 1/1 GPU 3 pool 1
 [2012-10-12 18:02:24] Accepted 122c1ff1 Diff 14/1 GPU 1 pool 1

The 8 byte hex value are the 2nd 8 bytes of the share being submitted to the
pool. The 2 diff values are the actual difficulty target that share reached
followed by the difficulty target the pool is currently asking for.

The output line shows the following:
(5s):1713.6 (avg):1707.8 Mh/s | A:729  R:8  HW:0  WU:22.53/m

Each column is as follows:
5s:  A 5 second exponentially decaying average hash rate
avg: An all time average hash rate
A:  The total difficulty of Accepted shares
R:  The total difficulty of Rejected shares
HW:  The number of HardWare errors
WU:  The Work Utility defined as the number of diff1 shares work / minute
     (accepted or rejected).

 GPU 1: 73.5C 2551RPM | 427.3/443.0Mh/s | A:8 R:0 HW:0 WU:4.39/m

Each column is as follows:
Temperature (if supported)
Fanspeed (if supported)
A 5 second exponentially decaying average hash rate
An all time average hash rate
The total difficulty of accepted shares
The total difficulty of rejected shares
The number of hardware erorrs
The work utility defined as the number of diff1 shares work / minute

The sgminer status line shows:
 ST: 1  SS: 0  NB: 1  LW: 8  GF: 1  RF: 1

ST is STaged work items (ready to use).
SS is Stale Shares discarded (detected and not submitted so don't count as rejects)
NB is New Blocks detected on the network
LW is Locally generated Work items
GF is Getwork Fail Occasions (server slow to provide work)
RF is Remote Fail occasions (server slow to accept work)

The block display shows:
Block: 0074c5e482e34a506d2a051a...  Started: [17:17:22]  Best share: 2.71K

This shows a short stretch of the current block, when the new block started,
and the all time best difficulty share you've found since starting sgminer
this time.


## Multipool

### Failover strategies

A number of different strategies for dealing with multipool setups are
available. Each has their advantages and disadvantages so multiple strategies
are available by user choice, as per the following list:

#### Failover

The default strategy is failover. This means that if you input a number of
pools, it will try to use them as a priority list, moving away from the 1st
to the 2nd, 2nd to 3rd and so on. If any of the earlier pools recover, it will
move back to the higher priority ones.

#### Round robin

This strategy only moves from one pool to the next when the current one falls
idle and makes no attempt to move otherwise.

#### Rotate

This strategy moves at user-defined intervals from one active pool to the next,
skipping pools that are idle.

#### Load balance

This strategy sends work to all the pools on a quota basis. By default, all
pools are allocated equal quotas unless specified with --quota. This
apportioning of work is based on work handed out, not shares returned so is
independent of difficulty targets or rejected shares. While a pool is disabled
or dead, its quota is dropped until it is re-enabled. Quotas are forward
looking, so if the quota is changed on the fly, it only affects future work.
If all pools are set to zero quota or all pools with quota are dead, it will
fall back to a failover mode. See quota below for more information.

The failover-only flag has special meaning in combination with load-balance
mode and it will distribute quota back to priority pool 0 from any pools that
are unable to provide work for any reason so as to maintain quota ratios
between the rest of the pools.

#### Balance

This strategy monitors the amount of difficulty 1 shares solved for each pool
and uses it to try to end up doing the same amount of work for all pools.


### Quotas

The load-balance multipool strategy works off a quota based scheduler. The
quotas handed out by default are equal, but the user is allowed to specify any
arbitrary ratio of quotas. For example, if all the quota values add up to 100,
each quota value will be a percentage, but if 2 pools are specified and pool0
is given a quota of 1 and pool1 is given a quota of 9, pool0 will get 10% of
the work and pool1 will get 90%. Quotas can be changed on the fly by the API,
and do not act retrospectively. Setting a quota to zero will effectively
disable that pool unless all other pools are disabled or dead. In that
scenario, load-balance falls back to regular failover priority-based strategy.
While a pool is dead, it loses its quota and no attempt is made to catch up
when it comes back to life.

To specify quotas on the command line, pools should be specified with a
semicolon separated --quota(or -U) entry instead of --url. Pools specified with
--url are given a nominal quota value of 1 and entries can be mixed.

For example:
--url poola:porta -u usernamea -p passa --quota "2;poolb:portb" -u usernameb -p passb
Will give poola 1/3 of the work and poolb 2/3 of the work.

Writing configuration files with quotas is likewise supported. To use
the above quotas in a configuration file they would be specified thus:

    "pools" : [
        {
                "url" : "poola:porta",
                "user" : "usernamea",
                "pass" : "passa"
        },
        {
                "quota" : "2;poolb:portb",
                "user" : "usernameb",
                "pass" : "passb"
        }
    ]


### Extra File Configuration

If you want to store a number of pools in your configuration file, but
don't always want them automatically enabled at start up (or restart),
then the "state" option with a value of "disabled" can be used:

    "pools" : [
        {
                "url" : "poola:porta",
                "user" : "usernamea",
                "pass" : "passa"
        },
        {
                "quota" : "2;poolb:portb",
                "user" : "usernameb",
                "pass" : "passb",
                "state" : "disabled"
        }
    ]

It is then trivial to change the "state" setting to "enabled" in the
configuration file at anytime and then restart the miner (see below).
You can enable the pool whilst the miner is still running ('p' followed
by 'e' followed by pool number) - but the pool will still be disabled on
restart if the config file is not changed.

"state" can also be set to "hidden". This allows the json file to
contain a large number of pools, of which some could be automatically
culled at start up. This makes it easy to swap pools in and out of the
runtime selection, without having a large list of pools cluttering up
the display.

    "pools" : [
        {
                "poolname" : "Main Pool",
                "url" : "poola:porta",
                "user" : "usernamea",
                "pass" : "passa",
                "state" : "disabled"
        },
        {
                "poolname" : "Joe's Weekend Pool",
                "quota" : "2;poolb:portb",
                "user" : "usernameb",
                "pass" : "passb",
                "state" : "hidden"
        }
    ]

These options are considered experimental and therefore will NOT be
created when the 'Write config file' option is used ('s' followed by
'w').

A restart of the miner ('s' followed by 'c') will reload the config
file and any changes that may have been made.


## Logging

sgminer will log to stderr if it detects stderr is being redirected to a
file. To enable logging simply append `2>logfile.txt` to your command line
and `logfile.txt` will contain all debug output unless you set `debug-log`
to `false`, in which case it will only contain output at the log level you
specified (notice by default).

There is also the -m option on Linux which will spawn a command of your choice
and pipe the output directly to that command.

The WorkTime details 'debug' option adds details on the end of each line
displayed for Accepted or Rejected work done. An example would be:

 <-00000059.ed4834a3 M:X D:1.0 G:17:02:38:0.405 C:1.855 (2.995) W:3.440 (0.000) S:0.461 R:17:02:47

The first 2 hex codes are the previous block hash, the rest are reported in
seconds unless stated otherwise:
The previous hash is followed by the getwork mode used M:X where X is one of
P:Pool, T:Test Pool, L:LP or B:Benchmark,
then D:d.ddd is the difficulty required to get a share from the work,
then G:hh:mm:ss:n.nnn, which is when the getwork or LP was sent to the pool and
the n.nnn is how long it took to reply,
followed by 'O' on it's own if it is an original getwork, or 'C:n.nnn' if it was
a clone with n.nnn stating how long after the work was recieved that it was cloned,
(m.mmm) is how long from when the original work was received until work started,
W:n.nnn is how long the work took to process until it was ready to submit,
(m.mmm) is how long from ready to submit to actually doing the submit, this is
usually 0.000 unless there was a problem with submitting the work,
S:n.nnn is how long it took to submit the completed work and await the reply,
R:hh:mm:ss is the actual time the work submit reply was received

If you start sgminer with the --sharelog option, you can get detailed
information for each share found. The argument to the option may be "-" for
standard output (not advisable with the ncurses UI), any valid positive number
for that file descriptor, or a filename.

To log share data to a file named "share.log", you can use either:
./sgminer --sharelog 50 -o xxx -u yyy -p zzz 50>share.log
./sgminer --sharelog share.log -o xxx -u yyy -p zzz

For every share found, data will be logged in a CSV (Comma Separated Value)
format:
    timestamp,disposition,target,pool,dev,thr,sharehash,sharedata
For example (this is wrapped, but it's all on one line for real):
    1335313090,reject,
    ffffffffffffffffffffffffffffffffffffffffffffffffffffffff00000000,
    http://localhost:8337,GPU0,0,
    6f983c918f3299b58febf95ec4d0c7094ed634bc13754553ec34fc3800000000,
    00000001a0980aff4ce4a96d53f4b89a2d5f0e765c978640fe24372a000001c5
    000000004a4366808f81d44f26df3d69d7dc4b3473385930462d9ab707b50498
    f681634a4f1f63d01a0cd43fb338000000000080000000000000000000000000
    0000000000000000000000000000000000000000000000000000000080020000


NEWS.md
# Release news

## Development 5.1.0-dev

* Added support for neoscrypt (by _ystarnaud_)
* Implemented the kernel updates for Xn algorithms, groestlcoin and others (by _badman74_)
* Added support for diamond v2 and whirlcoin (by _badman74_)
* Added a new scrypt kernel ("arebyp") (by _badman74_)

## Master

* Added support for animecoin, darkcoin, fuguecoin, groestlcoin, inkcoin,
  marucoin, myriadcoin-groestl, quarkcoin, qubitcoin, sifcoin, twecoin,
  darkcoin-mod ("X11-mod"), marucoin-mod ("X13-mod"), maxcoin (by
  _mrbrdo_).
* `intensity`, `xintensity`, `rawintensity`, `gpu-memclock`,
  `gpu-engine`, `thread-concurrency`, `gpu-threads` now also have a
  `pool-*` version to configure them for each pool separately
  (by _mrbrdo_).
* Initial configuration system revamping (by _ystarnaud_).
* Algorithm profile configuration (by _ystarnaud_).
* Complete configuration documentation, see `doc/configuration.md` (by
  _ystarnaud_).
* API documentation update, see `doc/API.md` (by _ystarnaud_).
* Extranonce support for stratum (by _bitbandi_).


## Version 4.2.2 - 27th June 2014

* Fixes for a few stratum-related security vulnerabilities (reported by
  _Mick Ayzenberg_ of DejaVu Security).
* Found blocks calculation fix (by _troky_).


## Version 4.2.1 - 22nd May 2014

* Fixed MSVS building, tested with MSVC++ 2010 and 2013 (by _troky_).
* Added the "ultratune" feature from `sph-sgminer`, available in the
  NCurses interface with `[G][C][U]` (by _ultracorp_).


## Version 4.2.0 - 20th May 2014

* git repo moved to [sgminer-dev](https://github.com/sgminer-dev/sgminer).
  It is now a GitHub organisation with several people having write access.
* Kernel `bufius` merger from `vertminer` (by _Bufius_).
* Set pool as idle on several stratum failure conditions (by _elbandi_).
* API response to `version` has field `CGMiner` instead of `SGMiner`
  for API client compatibility, and an additional `Miner` field (by
  _luke-jr_).
* API response to `restart` and `quit` only contains a `status` section
  and passes JSON validation (by _luke-jr_).
* API response to `devs` contains `XIntensity` and `RawIntensity` fields.
* Config file writing from TUI/API should produce a borked config less
  often.


## Version 4.1.271 - 12th April 2014

* Allow setting algorithm per-pool and initial implementation of kernel
  hot-swapping (by _mrbrdo_). Use options `pool-algorithm` (in config
  file or in CLI) or `algorithm` (config-only).
* NCurses UI `[S][W]` writes pool name and description to configuration.
* Added algorithm name to `pools` API command (by _troky_).
* NCurses UI asks for (optional) pool name, description and algorithm
  when adding pool.
* API command `addpool` allows both `url,user,pass` and
  `url,user,pass,name,desc,algo`.


## Version 4.1.242 - 7th April 2014

* There are now two mailing lists (on SourceForge), see `README.md`.
* Adaptive-N-factor algorithm support (by _Bufius_, _Zuikkis_ and
  _veox_). For details on choosing the algorithm, see
  `doc/configuration.md`.
* Allowed kernel names are no longer hard-coded. It is now possible to
  use any `.cl` file.
* Configuration parameter `poolname` has been renamed to `name`.
  `poolname` is deprecated and will be removed in a future version.
* Multiple `--name` parsing should now work as expected (by _troky_).
* `--description` configuration parameter to specify a freeform pool
  description, and `--priority` to specify the pool's priority (by
  _troky_).


## Version 4.1.153 - 14th March 2014

* Display pool URL instead of "Pool N" if no `poolname` specified.
* Incognito mode to hide user name in NCurses interface - useful for
  publishing screenshots. To enable, use `--incognito` on command line,
  `incognito` in config or `[D][I]` in NCurses interface.
* Support building in Microsoft Visual Studio 2010, perhaps other
  versions as well (by _troky_). Documentation in `winbuild/README.txt`.
* Support building in Cygwin (by _markuspeloquin_). Documentation in
  `doc/cygwin-build.txt`.
* Forward-port changes from `ckolivas/cgminer` up to 3.12.3.
* Allow setting `worksize` for kernel `zuikkis`.
* More log messages in pool handling.
* Updated `doc/FAQ.md`.
* Updated `example.conf`.


## Version 4.1.0 - 7th February 2014

* Writing configuration file from NCurses interface is broken!
* Commandline option parsing may be broken!
* MinGW building may be broken!
* Bug reporting documentation: `doc/BUGS.md`.
* Kernel selection and integration: `doc/KERNEL.md`.
* Several updates to other documentation files.
* Warn if `GPU_*` environment variables are not set.
* Maximum `intensity` lowered to 31 (anything above that gives an overflow
  anyway).
* Experimental `xintensity` setting (by _Kalroth_), see
  [commit message](https://github.com/veox/sgminer/commit/7aeae40af22e6108aab8b68a229eea25a639d650).
* Experimental `rawintensity` setting (by _Kalroth_), see
  [commit message](https://github.com/veox/sgminer/commit/d11df698d141988491494aa1f29c7d3595e9712b).
* `-v` is now a shorthand for `--verbose`, not `--vectors`.
* Default `scantime` and `expiry` changed to 7 and 28 (tests run by
  _MissedOutOnBTC_).
* Increased precision of `API_MHS`; added `API_KHS`.
* Pulled in kernels: `alexkarold`, `alexkarnew`, `psw`, `zuikkis`; renamed
  `scrypt` to `ckolivas`; all kernels now reside in directory `kernel`.
  Kernels can be chosen at startup only, by specifying `kernel`.
* Small optimisation to `ckolivas` kernel (by _gdevenyi_).
* Named pools via `poolname` (by _Kalroth_).
* Failover pool switching back delay is configurable via
  `failover-switch-delay` (by _Kalroth_).
* Pool `state`: `enabled`, `disabled`, and `hidden` (by _Joe4782_).
* Allow all pools to be set `disabled`.
* Use RPM in ADL `get-fanspeed` requests (from `bfgminer`, by _luke-jr_).
* Verbose ADL failure messages (by _Joe4782_ and _deba12_).
* Use `git` version string if available.
* Allow bypassing ADL checks during build with `--disable-adl-checks`.
* MinGW build checks (by _tonobitc_).

api.h
#ifndef API_H
#define API_H

#include "config.h"

#include "miner.h"

// BUFSIZ varies on Windows and Linux
#define TMPBUFSIZ 8192

// Number of requests to queue - normally would be small
#define QUEUE 100

#define COMSTR ","
#define SEPSTR "|"

#define CMDJOIN '+'
#define JOIN_CMD "CMD="
#define BETWEEN_JOIN SEPSTR
#define _DYNAMIC "D"

#define _DEVS   "DEVS"
#define _POOLS    "POOLS"
#define _PROFILES    "PROFILES"
#define _SUMMARY  "SUMMARY"
#define _STATUS   "STATUS"
#define _VERSION  "VERSION"
#define _MINECONFIG "CONFIG"
#define _GPU    "GPU"

#define _GPUS   "GPUS"
#define _NOTIFY   "NOTIFY"
#define _DEVDETAILS "DEVDETAILS"
#define _BYE    "BYE"
#define _RESTART  "RESTART"
#define _MINESTATS  "STATS"
#define _CHECK    "CHECK"
#define _MINECOIN "COIN"
#define _DEBUGSET "DEBUG"
#define _SETCONFIG  "SETCONFIG"

#define JSON0   "{"
#define JSON1   "\""
#define JSON2   "\":["
#define JSON3   "]"
#define JSON4   ",\"id\":1"
// If anyone cares, id=0 for truncated output
#define JSON4_TRUNCATED ",\"id\":0"
#define JSON5   "}"

#define JSON_START  JSON0
#define JSON_DEVS JSON1 _DEVS JSON2
#define JSON_POOLS  JSON1 _POOLS JSON2
#define JSON_PROFILES  JSON1 _POOLS JSON2
#define JSON_SUMMARY  JSON1 _SUMMARY JSON2
#define JSON_STATUS JSON1 _STATUS JSON2
#define JSON_VERSION  JSON1 _VERSION JSON2
#define JSON_MINECONFIG JSON1 _MINECONFIG JSON2
#define JSON_GPU  JSON1 _GPU JSON2

#define JSON_GPUS JSON1 _GPUS JSON2
#define JSON_NOTIFY JSON1 _NOTIFY JSON2
#define JSON_DEVDETAILS JSON1 _DEVDETAILS JSON2
#define JSON_CLOSE  JSON3
#define JSON_MINESTATS  JSON1 _MINESTATS JSON2
#define JSON_CHECK  JSON1 _CHECK JSON2
#define JSON_MINECOIN JSON1 _MINECOIN JSON2
#define JSON_DEBUGSET JSON1 _DEBUGSET JSON2
#define JSON_SETCONFIG  JSON1 _SETCONFIG JSON2

#define JSON_END  JSON4 JSON5
#define JSON_END_TRUNCATED  JSON4_TRUNCATED JSON5
#define JSON_BETWEEN_JOIN ","

#define MSG_INVGPU 1
#define MSG_ALRENA 2
#define MSG_ALRDIS 3
#define MSG_GPUMRE 4
#define MSG_GPUREN 5
#define MSG_GPUNON 6
#define MSG_POOL 7
#define MSG_NOPOOL 8
#define MSG_DEVS 9
#define MSG_NODEVS 10
#define MSG_SUMM 11
#define MSG_GPUDIS 12
#define MSG_GPUREI 13
#define MSG_INVCMD 14
#define MSG_MISID 15
#define MSG_GPUDEV 17

#define MSG_NUMGPU 20

#define MSG_VERSION 22
#define MSG_INVJSON 23
#define MSG_MISCMD 24
#define MSG_MISPID 25
#define MSG_INVPID 26
#define MSG_SWITCHP 27
#define MSG_MISVAL 28
#define MSG_NOADL 29
#define MSG_NOGPUADL 30
#define MSG_INVINT 31
#define MSG_GPUINT 32
#define MSG_MINECONFIG 33
#define MSG_GPUMERR 34
#define MSG_GPUMEM 35
#define MSG_GPUEERR 36
#define MSG_GPUENG 37
#define MSG_GPUVERR 38
#define MSG_GPUVDDC 39
#define MSG_GPUFERR 40
#define MSG_GPUFAN 41
#define MSG_MISFN 42
#define MSG_BADFN 43
#define MSG_SAVED 44
#define MSG_ACCDENY 45
#define MSG_ACCOK 46
#define MSG_ENAPOOL 47
#define MSG_DISPOOL 48
#define MSG_ALRENAP 49
#define MSG_ALRDISP 50
#define MSG_MISPDP 52
#define MSG_INVPDP 53
#define MSG_TOOMANYP 54
#define MSG_ADDPOOL 55

#define MSG_NOTIFY 60

#define MSG_REMLASTP 66
#define MSG_ACTPOOL 67
#define MSG_REMPOOL 68
#define MSG_DEVDETAILS 69
#define MSG_MINESTATS 70
#define MSG_MISCHK 71
#define MSG_CHECK 72
#define MSG_POOLPRIO 73
#define MSG_DUPPID 74
#define MSG_MISBOOL 75
#define MSG_INVBOOL 76
#define MSG_FOO 77
#define MSG_MINECOIN 78
#define MSG_DEBUGSET 79
#define MSG_SETCONFIG 82
#define MSG_UNKCON 83
#define MSG_INVNUM 84
#define MSG_CONPAR 85
#define MSG_CONVAL 86

#define MSG_NOUSTA 88

#define MSG_ZERMIS 94
#define MSG_ZERINV 95
#define MSG_ZERSUM 96
#define MSG_ZERNOSUM 97

#define MSG_BYE 0x101

#define MSG_INVNEG 121
#define MSG_SETQUOTA 122
#define MSG_LOCKOK 123
#define MSG_LOCKDIS 124

#define MSG_CHSTRAT 125
#define MSG_MISSTRAT 126
#define MSG_INVSTRAT 127
#define MSG_MISSTRATINT 128

#define MSG_PROFILE 129
#define MSG_NOPROFILE 130

#define MSG_PROFILEEXIST 131
#define MSG_MISPRD 132
#define MSG_ADDPROFILE 133

#define MSG_MISPRID 134
#define MSG_PRNOEXIST 135
#define MSG_PRISDEFAULT 136
#define MSG_PRINUSE 137
#define MSG_REMPROFILE 138

#define MSG_CHPOOLPR 139

#define MSG_INVXINT 140
#define MSG_GPUXINT 141
#define MSG_INVRAWINT 142
#define MSG_GPURAWINT 143

enum code_severity {
  SEVERITY_ERR,
  SEVERITY_WARN,
  SEVERITY_INFO,
  SEVERITY_SUCC,
  SEVERITY_FAIL
};

enum code_parameters {
  PARAM_GPU,
  PARAM_PID,
  PARAM_GPUMAX,
  PARAM_PMAX,
  PARAM_PRMAX,
  PARAM_POOLMAX,

// Single generic case: have the code resolve it - see below
  PARAM_DMAX,

  PARAM_CMD,
  PARAM_POOL,
  PARAM_STR,
  PARAM_BOTH,
  PARAM_BOOL,
  PARAM_SET,
  PARAM_INT,
  PARAM_NONE
};

struct CODES {
  const enum code_severity severity;
  const int code;
  const enum code_parameters params;
  const char *description;
};

extern struct CODES codes[];

struct IP4ACCESS {
  in_addr_t ip;
  in_addr_t mask;
  char group;
};

#define GROUP(g) (toupper(g))
#define PRIVGROUP GROUP('W')
#define NOPRIVGROUP GROUP('R')
#define ISPRIVGROUP(g) (GROUP(g) == PRIVGROUP)
#define GROUPOFFSET(g) (GROUP(g) - GROUP('A'))
#define VALIDGROUP(g) (GROUP(g) >= GROUP('A') && GROUP(g) <= GROUP('Z'))
#define COMMANDS(g) (apigroups[GROUPOFFSET(g)].commands)
#define DEFINEDGROUP(g) (ISPRIVGROUP(g) || COMMANDS(g) != NULL)

struct io_data {
  size_t siz;
  char *ptr;
  char *cur;
  bool sock;
  bool close;
};

struct io_list {
  struct io_data *io_data;
  struct io_list *prev;
  struct io_list *next;
};

extern void message(struct io_data *io_data, int messageid, int paramid, char *param2, bool isjson);
extern bool io_add(struct io_data *io_data, char *buf);
extern void io_close(struct io_data *io_data);
extern void io_free();

extern struct api_data *api_add_escape(struct api_data *root, char *name, char *data, bool copy_data);
extern struct api_data *api_add_string(struct api_data *root, char *name, char *data, bool copy_data);
extern struct api_data *api_add_const(struct api_data *root, char *name, const char *data, bool copy_data);
extern struct api_data *api_add_uint8(struct api_data *root, char *name, uint8_t *data, bool copy_data);
extern struct api_data *api_add_uint16(struct api_data *root, char *name, uint16_t *data, bool copy_data);
extern struct api_data *api_add_int(struct api_data *root, char *name, int *data, bool copy_data);
extern struct api_data *api_add_uint(struct api_data *root, char *name, unsigned int *data, bool copy_data);
extern struct api_data *api_add_uint32(struct api_data *root, char *name, uint32_t *data, bool copy_data);
extern struct api_data *api_add_hex32(struct api_data *root, char *name, uint32_t *data, bool copy_data);
extern struct api_data *api_add_uint64(struct api_data *root, char *name, uint64_t *data, bool copy_data);
extern struct api_data *api_add_double(struct api_data *root, char *name, double *data, bool copy_data);
extern struct api_data *api_add_elapsed(struct api_data *root, char *name, double *data, bool copy_data);
extern struct api_data *api_add_bool(struct api_data *root, char *name, bool *data, bool copy_data);
extern struct api_data *api_add_timeval(struct api_data *root, char *name, struct timeval *data, bool copy_data);
extern struct api_data *api_add_time(struct api_data *root, char *name, time_t *data, bool copy_data);
extern struct api_data *api_add_mhs(struct api_data *root, char *name, double *data, bool copy_data);
extern struct api_data *api_add_khs(struct api_data *root, char *name, double *data, bool copy_data);
extern struct api_data *api_add_mhtotal(struct api_data *root, char *name, double *data, bool copy_data);
extern struct api_data *api_add_temp(struct api_data *root, char *name, float *data, bool copy_data);
extern struct api_data *api_add_utility(struct api_data *root, char *name, double *data, bool copy_data);
extern struct api_data *api_add_freq(struct api_data *root, char *name, double *data, bool copy_data);
extern struct api_data *api_add_volts(struct api_data *root, char *name, float *data, bool copy_data);
extern struct api_data *api_add_hs(struct api_data *root, char *name, double *data, bool copy_data);
extern struct api_data *api_add_diff(struct api_data *root, char *name, double *data, bool copy_data);
extern struct api_data *api_add_percent(struct api_data *root, char *name, double *data, bool copy_data);
extern struct api_data *api_add_avg(struct api_data *root, char *name, float *data, bool copy_data);
extern struct api_data *print_data(struct api_data *root, char *buf, bool isjson, bool precom);

#define SOCKBUFALLOCSIZ 65536

#define io_new(init) _io_new(init, false)
#define sock_io_new() _io_new(SOCKBUFALLOCSIZ, true)

#if LOCK_TRACKING

  #define LOCK_FMT_FFL " - called from %s %s():%d"
  #define LOCKMSG(fmt, ...) fprintf(stderr, "APILOCK: " fmt "\n", ##__VA_ARGS__)
  #define LOCKMSGMORE(fmt, ...) fprintf(stderr, "          " fmt "\n", ##__VA_ARGS__)
  #define LOCKMSGFFL(fmt, ...) fprintf(stderr, "APILOCK: " fmt LOCK_FMT_FFL "\n", ##__VA_ARGS__, file, func, linenum)
  #define LOCKMSGFLUSH() fflush(stderr)

  typedef struct lockstat {
    uint64_t lock_id;
    const char *file;
    const char *func;
    int linenum;
    struct timeval tv;
  } LOCKSTAT;

  typedef struct lockline {
    struct lockline *prev;
    struct lockstat *stat;
    struct lockline *next;
  } LOCKLINE;

  typedef struct lockinfo {
    void *lock;
    enum cglock_typ typ;
    const char *file;
    const char *func;
    int linenum;
    uint64_t gets;
    uint64_t gots;
    uint64_t tries;
    uint64_t dids;
    uint64_t didnts; // should be tries - dids
    uint64_t unlocks;
    LOCKSTAT lastgot;
    LOCKLINE *lockgets;
    LOCKLINE *locktries;
  } LOCKINFO;

  typedef struct locklist {
    LOCKINFO *info;
    struct locklist *next;
  } LOCKLIST;
#endif

#endif /* API_H */